# TIDY-NCBI-GENOME-DOWNLOAD

[![codecov](https://codecov.io/gh/snail123815/tidy-ncbi-genome-download/branch/main/graph/badge.svg?token=1MHRVD759X)](https://codecov.io/gh/snail123815/tidy-ncbi-genome-download)

This is a tool set dealing with downloaded genomes from NCBI using [ncbi-genome-download](https://github.com/kblin/ncbi-genome-download).

The purpose of these tools is to connect *semi-automatically* to downstream tools. For example [phylophlan](https://github.com/biobakery/phylophlan) and [mashtree](https://github.com/biobakery/phylophlan), which needs a set of genome files in one folder, uses the file names in their output, and raises error if there are unaccepted characters (eg. `(` or space) in file names.

## Download target genomes from NCBI.

```shell
ncbi-genome-download -F genbank -g "Streptomyces,Kitasatospora" -H -p 10 -r 3 -m ncbiftp-Streptomyces-Kitasatospora-gbk.tsv -o ncbiftp-Streptomyces-Kitasatospora-gbk bacteria
```

Support download using `-F genbank/fasta/protein-fasta`. The `-m` switch (`--metadata-table`) is required. The `-H` switch is optional,

## `gather_assemblies.py`

Note please skip `--maxCtg` when protein fasta files are downloaded.

```
usage: gather_assemblies.py [-h] [--excludeList EXCLUDELIST] [--maxCtg MAXCTG] [--targetDir TARGETDIR] tsv dir

positional arguments:
  tsv                   Path to the .tsv file generated by `-m` switch
  dir                   Path to the directory generated by `-o` parameter

options:
  -h, --help            show this help message and exit
  --excludeList EXCLUDELIST
                        Exclusion list file, one item per line
  --maxCtg MAXCTG       Maximum number of contigs that a genome will be kept.
  --targetDir TARGETDIR
                        Valid assemblies will be copied to this directory.
```

## `check_combine.py` and `combine_database.py`

Check file names validity if we want to combine from other sources.

```
usage: check_combine.py [-h] p [p ...]

positional arguments:
  p           pathes of databases (folders) you want to combine

options:
  -h, --help   show this help message and exit
```

Do the actual combining.

```
usage: combine_database.py [-h] [-t T] [--keep KEEP] p [p ...]

positional arguments:
  p            pathes of databases (folders) you want to combine

options:
  -h, --help   show this help message and exit
  -t T         target dir to store combined files
  --keep KEEP  If duplicated file names found, keep "first" or "all"
```
